src/shogun/mathematics/linalg/LinalgEnums.h

```diff
@@ -32,12 +32,20 @@
 #ifndef LINALG_ENUMS_H__
 #define LINALG_ENUMS_H__
 
+#include <type_traits>
+#include <shogun/lib/common.h>
+
 namespace shogun
 {
 
 	namespace linalg
 	{
-
+		template <typename T, typename U>
+		struct promote
+		{
+			using type = decltype(T(0) + U(0));
```

Comment:
can you not just do decltype(T + U);?

Reply:
nvm!

Reply:
actually to generalise this you could do `decltype(T{}+U{})`

Reply:
this is much cleaner